### PR TITLE
OHRM5X-1225 : Preserve and display purged employee timesheet comments

### PR DESCRIPTION
**Pull request by SharukaNP**
Preserving the purged employee comments and actions made on other employee timesheets.

Displaying them in timesheets.

### CODE REVIEW

File: symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php [hunk 1]
```diff
@@ -19,43 +19,55 @@
 
 namespace OrangeHRM\Time\Api\Model;
 
+use DoctrineProxies\__CG__\OrangeHRM\Entity\EmployeeTerminationRecord;
```
Comment (devishke-orange) [inline symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php:22]: Check this import must've accidentally selected this in IDE

File: installer/dbscript/dbscript-2.sql [hunk 2]
```diff
@@ -4610,3 +4610,9 @@ SET @auth_admin_verify_data_group_id := (SELECT `id` FROM ohrm_data_group WHERE
 
 INSERT INTO ohrm_user_role_data_group (`can_read`, `can_create`, `can_update`, `can_delete`, `self`, `data_group_id`,`user_role_id`)
 VALUES (1, 0, 0, 0, 0, @auth_admin_verify_data_group_id, @admin_role_id);
+
+ALTER TABLE `ohrm_timesheet_action_log` CHANGE `performed_by` `performed_by` INT(20) NULL DEFAULT NULL;
+
+ALTER TABLE `ohrm_timesheet_action_log` DROP FOREIGN KEY `ohrm_timesheet_action_log_ibfk_1`;
+
+ALTER TABLE `ohrm_timesheet_action_log` ADD CONSTRAINT `ohrm_timesheet_action_log_performed_by_id` FOREIGN KEY (`performed_by`) REFERENCES `ohrm_user`(`id`) ON DELETE SET NULL ON UPDATE RESTRICT;
```
Comment (devishke-orange) [inline installer/dbscript/dbscript-2.sql:4618]: should we include `ON UPDATE RESTRICT` @RajithaKumara ? we got this query from phpmyadmin when changing the table

**Review by devishke-orange — CHANGES_REQUESTED** [review]
Looks good 👍

File: symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php (hunk 1 above)
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php:22]: what is the usage of this import?

File: symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php [hunk 3]
```diff
@@ -19,43 +19,55 @@
...
 use OrangeHRM\Core\Api\V2\Serializer\ModelTrait;
 use OrangeHRM\Core\Api\V2\Serializer\Normalizable;
 use OrangeHRM\Entity\TimesheetActionLog;
+use OrangeHRM\Entity\User;
 
 class TimesheetActionLogModel implements Normalizable
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php:28]: Do we need to change this model after you make the `performedUser` nullable in the entity class?

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php (hunk 1 above)
Comment (SharukaNP) [reply symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php:22]: This should be changed to the `use OrangeHRM\Entity\EmployeeTerminationRecord;`

File: symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php (hunk 3 above)
Comment (SharukaNP) [reply symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php:28]: User Entity does not have a direct connection with Timesheet so changing the TimesheetActionLog Model and Entity should be enough right?

File: installer/dbscript/dbscript-2.sql (hunk 2 above)
Comment (RajithaKumara) [reply installer/dbscript/dbscript-2.sql:4618]: Better to check DB compatibility with suported min max versions. e.g. MySQL 5.5 & MySQL 5.7, MariaDB 5.5 & MariaDB 10.6

File: symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php (hunk 3 above)
Comment (SharukaNP) [reply symfony/plugins/orangehrmTimePlugin/Api/Model/TimesheetActionLogModel.php:28]: Set the model back to the original. Made a small change in blade instead to handle the null value in performedBy

File: installer/dbscript/dbscript-2.sql (hunk 2 above)
Comment (SharukaNP) [reply installer/dbscript/dbscript-2.sql:4618]: MySQL 5.5 have support http://download.nust.na/pub6/mysql/doc/refman/5.5/en/innodb-foreign-key-constraints.html
MySQL 5.7 have https://dev.mysql.com/doc/refman/5.7/en/create-table-foreign-keys.html
MariaDB 5.5 have https://mariadb.com/kb/en/constraint/
MariaDB 10.6 have https://mariadb.com/kb/en/foreign-keys/

File: installer/dbscript/dbscript-2.sql [hunk 4]
```diff
@@ -4610,3 +4610,9 @@ SET @auth_admin_verify_data_group_id := (SELECT `id` FROM ohrm_data_group WHERE
 
 INSERT INTO ohrm_user_role_data_group (`can_read`, `can_create`, `can_update`, `can_delete`, `self`, `data_group_id`,`user_role_id`)
 VALUES (1, 0, 0, 0, 0, @auth_admin_verify_data_group_id, @admin_role_id);
+
+ALTER TABLE `ohrm_timesheet_action_log` CHANGE `performed_by` `performed_by` INT(20) NULL DEFAULT NULL;
+
+ALTER TABLE `ohrm_timesheet_action_log` DROP FOREIGN KEY `ohrm_timesheet_action_log_ibfk_1`;
```
Comment (RajithaKumara) [inline installer/dbscript/dbscript-2.sql:4616]: check this `ohrm_timesheet_action_log_ibfk_1`

File: symfony/client/src/orangehrmTimePlugin/components/TimesheetActions.vue [hunk 5]
```diff
@@ -57,9 +57,8 @@ const actionsNormalizer = data => {
       action: item.action?.label,
       date: item.date,
       comment: item.comment,
-      performedBy: `${item.performedEmployee?.firstName} ${
-        item.performedEmployee?.lastName
-      } 
+      performedBy: `${item.performedEmployee?.firstName ? item.performedEmployee?.firstName : 'Purged'}
+      ${ item.performedEmployee?.lastName ? item.performedEmployee?.lastName : 'Empployee' }
```
Comment (RajithaKumara) [inline symfony/client/src/orangehrmTimePlugin/components/TimesheetActions.vue:61]:
```suggestion
      ${ item.performedEmployee?.lastName ? item.performedEmployee?.lastName : 'Employee' }
```
??

File: symfony/plugins/orangehrmAttendancePlugin/Dao/AttendanceDao.php [hunk 6]
```diff
@@ -275,7 +275,13 @@ public function deleteAttendanceRecords(array $attendanceRecordIds): int
     public function getAttendanceRecordById(int $attendanceRecordId): ?AttendanceRecord
     {
         $attendanceRecord = $this->getRepository(AttendanceRecord::class)->find($attendanceRecordId);
-        return ($attendanceRecord instanceof AttendanceRecord) ? $attendanceRecord : null;
+        if ($attendanceRecord instanceof AttendanceRecord) {
+            if (!is_null($attendanceRecord->getEmployee()->getPurgedAt())) {
+                return null;
+            }
+            return $attendanceRecord;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAttendancePlugin/Dao/AttendanceDao.php:282]:
```suggestion
            if (is_null($attendanceRecord->getEmployee()->getPurgedAt())) {
                return $attendanceRecord;
            }
```

File: symfony/plugins/orangehrmTimePlugin/entity/TimesheetActionLog.php [hunk 7]
```diff
@@ -145,9 +146,9 @@ public function setDate(DateTime $date): void
     }
 
     /**
-     * @return User
+     * @return User|null
      */
-    public function getPerformedUser(): User
+    public function getPerformedUser(): ?User
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/entity/TimesheetActionLog.php:151]: check `setPerformedUser` as well

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/client/src/orangehrmTimePlugin/components/TimesheetActions.vue (hunk 5 above)
Comment (SharukaNP) [reply symfony/client/src/orangehrmTimePlugin/components/TimesheetActions.vue:61]: changed

File: symfony/plugins/orangehrmTimePlugin/entity/TimesheetActionLog.php (hunk 7 above)
Comment (SharukaNP) [reply symfony/plugins/orangehrmTimePlugin/entity/TimesheetActionLog.php:151]: changed the seterformedUser

File: installer/dbscript/dbscript-2.sql (hunk 4 above)
Comment (SharukaNP) [reply installer/dbscript/dbscript-2.sql:4616]: the default fk generated seems to be the same even when I did a fresh install

File: symfony/plugins/orangehrmAttendancePlugin/Dao/AttendanceDao.php (hunk 6 above)
Comment (SharukaNP) [reply symfony/plugins/orangehrmAttendancePlugin/Dao/AttendanceDao.php]: Changed

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**